diff --git a/Modules/posixmodule.c b/Modules/posixmodule.c
--- a/Modules/posixmodule.c
+++ b/Modules/posixmodule.c
@@ -192,6 +192,12 @@
 #  include <linux/memfd.h>        // memfd_create(), MFD_CLOEXEC
 #endif
 
+#ifdef HAVE_MEMFD_CREATE
+/* Weak references. */
+__attribute__((weak))
+int memfd_create(const char *name, unsigned int flags);
+#endif
+
 #ifdef HAVE_SYS_EVENTFD_H
 #  include <sys/eventfd.h>        // eventfd()
 #endif
@@ -18128,6 +18134,18 @@ posixmodule_exec(PyObject *m)
 {
     _posixstate *state = get_posix_state(m);
 
+#ifdef HAVE_MEMFD_CREATE
+    if (memfd_create == NULL) {
+        PyObject *dict = PyModule_GetDict(m);
+        if (dict == NULL) {
+            return -1;
+        }
+        if (PyDict_PopString(dict, "memfd_create", NULL) < 0) {
+            return -1;
+        }
+    }
+#endif
+
 #if defined(HAVE_PWRITEV)
     if (HAVE_PWRITEV_RUNTIME) {} else {
         PyObject* dct = PyModule_GetDict(m);
